            

                                                                Module-3 : Pandas
                                                      Submitted By:Seela Sri Lakshmi Gayathri
                                                                Roll No:238T1A12A5
                                                                  Batch:03

# Section - A : Output Prediction

### Q1.  Array operations and vectorization

import numpy as np

import pandas as pd

arr = np.array([10, 20, 30])

s = pd.Series([10, 20, 30])

print(arr * 2)

print(s * 2)

#### ◆ Predict the output of both statements.

[20  40  60]

0       20

1       40

2       60

dtype: int64

#### ◆ Identify one similarity and one difference between the two data structures

Similarity:Both perform element-wise multiplication by 2.

Difference: NumPy array displays only the values, while Pandas Series displays values along with an index

In [2]:
import numpy as np

import pandas as pd

arr = np.array([10, 20, 30])

s = pd.Series([10, 20, 30])

print(arr * 2)

print(s * 2)

[20 40 60]
0    20
1    40
2    60
dtype: int64


### Q2. Boolean Filtering

import pandas as pd

df = pd.DataFrame({

    'Student': ['A', 'B', 'C', 'D'],

    'Marks': [45, 80, 65, 90]
})

result = df[df['Marks'] >= 65]

print(result[['Student', 'Marks']])

#### ◆ Predict the rows displayed.

   Student  Marks
   
1       B     80

2       C     65

3       D     90

#### ◆ How many rows and columns will the resulting DataFrame contain?

Rows: 3

Column: 2

Explanation:The condition Marks >= 65 selects only students whose marks are 65 or above.



In [1]:

import pandas as pd

df = pd.DataFrame({
    'Student': ['A', 'B', 'C', 'D'],
    'Marks': [45, 80, 65, 90]
})

result = df[df['Marks'] >= 65]

print(result[['Student', 'Marks']])



  Student  Marks
1       B     80
2       C     65
3       D     90


### Q3.  Descriptive statistics

import pandas as pd

df = pd.DataFrame({

    'Marks': [10, 20, 30, 40, 50]
})

print(df['Marks'].mean())

print(df['Marks'].max())

print(df['Marks'].median())

#### ◆ Predict all three outputs.

30.0

50

30

Here,the mean and median are both 30.0.

#### ◆ Explain how the median differs from the mean.

● Mean means to calculate the average of all values in a dataset.

● Median means to finding the middle value of a numeric dataset after sorting the values in ascending order.





In [3]:
import pandas as pd

df = pd.DataFrame({

    'Marks': [10, 20, 30, 40, 50]
})

print(df['Marks'].mean())

print(df['Marks'].max())

print(df['Marks'].median())

30.0
50
30.0


### Q4.  Group-wise Analysis

import pandas as pd

df = pd.DataFrame({

    'Department': ['IT', 'HR', 'IT', 'HR'],

    'Salary': [40000, 30000, 60000, 50000]
})

print(df.groupby('Department')['Salary'].mean())


#### ◆ Predict the output.

Department

HR    40000.0

IT    50000.0

Name: Salary, dtype: float64

#### ◆ Explain how groupby() calculates the department-wise average.

Explanation: 

groupby() separates the data according to department and then mean()
calculates the average salary for each department.

● IT = (40000 + 60000) / 2 = 50000

● HR = (30000 + 50000) / 2 = 40000



In [6]:
import pandas as pd

df = pd.DataFrame({

    'Department': ['IT', 'HR', 'IT', 'HR'],
    
    'Salary': [40000, 30000, 60000, 50000]
})

print(df.groupby('Department')['Salary'].mean())

Department
HR    40000.0
IT    50000.0
Name: Salary, dtype: float64


### Q5.   Pandas Correlation - Positive

import pandas as pd

df = pd.DataFrame({

    'Hours': [1, 2, 3, 4],

    'Marks': [20, 40, 60, 80]
})

print(df['Hours'].corr(df['Marks']))

#### ◆ Predict the correlation coefficient.

The correlation coefficient is 1.0

#### ◆ Interpret the direction and strength of the relationship.

Explanation: 

The correlation is +1.0, showing a perfect positive linear relationship. As study hours increase, marks also increase proportionally.



In [7]:
import pandas as pd

df = pd.DataFrame({

    'Hours': [1, 2, 3, 4],
    
    'Marks': [20, 40, 60, 80]
})

print(df['Hours'].corr(df['Marks']))

1.0


### Q6.  Pandas Correlation - Negative

import pandas as pd

df = pd.DataFrame({

    'Price': [10, 20, 30, 40],

    'Demand': [100, 80, 60, 40]
})

print(df['Price'].corr(df['Demand']))

#### ◆ Predict the correlation coefficient.

The correlation coefficient is -1.0

#### ◆ Explain what the sign of the coefficient indicates about the relationship.

Explanation: 

The correlation is -1.0, showing a perfect negative linear relationship. As price increases, demand decreases.



In [8]:
import pandas as pd

df = pd.DataFrame({

    'Price': [10, 20, 30, 40],
    
    'Demand': [100, 80, 60, 40]
})

print(df['Price'].corr(df['Demand']))


-1.0


# Section - B : Code Correction


### Q1. JSON and CSV          

import pandas as pd

df = pd.read_csv("students.csv", header=None)

print(df)    

A CSV file contains a header row, but the following code incorrectly reads the first student's record as column names. Correct the code.

### Error:

`header=None` tells pandas that the CSV file has no header row, so the first student's record is treated as data rather than column names.  
  
### Correction:
 
Remove `header=None` so pandas uses the first row as column headers.

### Explanation:

header=0 tells Pandas that the first row contains column names.

In [30]:
import pandas as pd
df = pd.read_csv("student.csv")
print(df)

   student_id student_name        Date  Marks
0         101     Gayathri  01-10-2026     80
1         102         Renu  02-10-2026     70
2         103    Sravanthi  03-10-2026     60
3         104        Neelu  04-10-2026     90
4         105        Pooji  05-10-2026     50


### Q2. Data Analysis

import pandas as pd

df = pd.DataFrame({

    'Employee': ['A', 'B', 'C', 'D', 'E', 'F'],
    
    'Salary': [30000, 50000, 25000, 60000, 45000, 70000]
})

print(df.sort_values('Salary').head(5))                              

The objective is to display the five highest salaries from a DataFrame. Correct the code.

### Error: 

sort_values('Salary') sorts salaries from lowest to highest, so head(5) gives the five lowest salaries.

### Correction:

 Use sort_values('Salary', ascending=False).head(5).

### Explanation:

ascending=False sorts salaries from highest to lowest, and head(5) selects the first five records.

In [15]:
import pandas as pd
df = pd.DataFrame({
    'Employee': ['A', 'B', 'C', 'D', 'E', 'F'], 
    'Salary': [30000, 50000, 25000, 60000, 45000, 70000]
})
print(df.sort_values('Salary', ascending=False).head(5))

  Employee  Salary
5        F   70000
3        D   60000
1        B   50000
4        E   45000
0        A   30000


### Q3. Cleaning Data          

import pandas as pd

df = pd.DataFrame({'Marks': [70, None, 90]})

df['Marks'].fillna(df['Marks'].mean())

print(df)    

A student wants to replace missing marks with the column's mean but writes code that does not update the DataFrame. Correct the code.

### Error:

fillna() returns the updated values but does not modify the DataFrame by default.

### Correction:

Use df['Marks'] = df['Marks'].fillna(df['Marks'].mean()).

### Explanation:

The mean of the available marks is 80, so the missing value is replaced with 80 and the DataFrame is updated.



In [16]:

import pandas as pd
df = pd.DataFrame({'Marks': [70, None, 90]})
df['Marks'] = df['Marks'].fillna(df['Marks'].mean())
print(df)

   Marks
0   70.0
1   80.0
2   90.0


### Q4. Correlation

import pandas as pd

df = pd.DataFrame({

    'Hours': [1, 2, 3, 4],

    'Marks': [20, 40, 60, 80]
})

print(df.corr('Hours', 'Marks’))

The objective is to calculate the correlation between Hours and Marks. Correct the code.

### Error:

df.corr('Hours', 'Marks') is incorrect because corr() does not take column names as arguments.

### Correction: 

df['Hours'].corr(df['Marks'])

### Explanation:

The corr() method calculates the correlation between the two selected columns. Here, the result is 1.0, showing a perfect positive correlation.


In [1]:

import pandas as pd

df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})

print(df[['Hours', 'Marks']].corr())



       Hours  Marks
Hours    1.0    1.0
Marks    1.0    1.0


# Section - C : Edge Cases



### Q1. Two NumPy arrays have shapes (3, 1) and (1, 3). Predict the shape of their element-wise sum and explain how broadcasting works. Why might the result differ from adding two one-dimensional arrays?


In [18]:
import numpy as np

a = np.array([[10], [20], [30]])

b = np.array([[1, 2, 3]])

result = a + b

print(result)

print(result.shape)

[[11 12 13]
 [21 22 23]
 [31 32 33]]
(3, 3)



### 1.Predict the output of the given code.

[[11 12 13]

 [21 22 23]

 [31 32 33]]

(3, 3)

### 2.Identify the shapes of arrays a and b before addition.

● a has 3 rows and 1 column

● b has 1 row and 3 columns



### 3.Explain how NumPy broadcasting enables the addition of these arrays.

NumPy broadcasting expands the compatible dimensions to perform `element-wise addition`, producing a (3, 3) array.


### 4.What would happen if b were replaced with np.array([1, 2])? Explain the reason.

If b is replaced with np.array([1, 2]), NumPy uses broadcasting to match the shapes of a and b. The addition is possible, and the output shape is (3, 2).

Reason: NumPy broadcasts the array [1, 2] across all three rows of a.
 

### Q2. A CSV file contains commas inside quoted text fields, missing values, and dates in different formats. Explain how incorrect parsing can affect the resulting DataFrame and identify suitable read options or preprocessing steps.


In [28]:
import pandas as pd

df = pd.read_csv("students.csv")
print(df)
print(df.dtypes)

    ID      Name        Date  Marks
0  101  "Asha,R"  01-09-2026   80.0
1  102      Ravi  01-09-2026    NaN
2  103      Neha  03-09-2026    NaN
ID         int64
Name         str
Date         str
Marks    float64
dtype: object



### Predict how the quoted comma in "Asha, R" will be handled when the CSV file is read.

The quoted comma in "Asha, R" is handled correctly by `read_csv()` and remains part of the name.

### Identify the potential issues with the Date and Marks columns.

The Date column contains different date formats, so dates should be converted into a consistent datetime format. The intended day/month order must be confirmed because dates can be ambiguous.

### Rewrite the code to handle missing marks and convert the date column into a suitable date-time format.

The Marks column contains missing values. `pd.to_numeric(errors="coerce")` converts invalid or non-numeric values to `NaN`.

### Explain why the value "NA" should be investigated before deciding whether it represents a missing value.

The value `"NA"` should be investigated to determine whether it means a missing mark or represents another category.

### Q3. A salary column contains the values 25000, 30000, 35000, and 10000000. Explain how the extreme value affects the mean compared with the median. Identify which statistic may better represent a typical employee's salary and justify your choice.

In [2]:

import pandas as pd

df = pd.DataFrame({
    "Employee": ["A", "B", "C", "D"],
    "Salary": [25000, 30000, 35000, 10000000]
})

print("Mean:", df["Salary"].mean())
print("Median:", df["Salary"].median())

Mean: 2522500.0
Median: 32500.0


### Predict the mean and median salary.

Mean: 2522500.0

Median: 32500.0

### Identify the effect of the extreme salary on the mean.

The extreme salary of 10,000,000 increases the mean significantly. 

### Which statistic may better represent a typical employee's salary in this dataset? Justify your answer.

Median (₹32,500) better represents a typical employee's salary.

Justification:

The median is less affected by extreme values or outliers. Since most employees earn between ₹25,000 and ₹35,000, ₹32,500 is closer to a typical employee's salary.


### Write code to identify the employee record with the highest salary and explain why the record should be investigated before removing it.



In [3]:
import pandas as pd
df = pd.DataFrame({
    "Employee": ["A", "B", "C", "D"],
    "Salary": [25000, 30000, 35000, 10000000]
})
print("Mean:", df["Salary"].mean())
print("Median:", df["Salary"].median())
highest_salary = df.loc[df["Salary"].idxmax()]
print("\nEmployee with highest salary:")
print(highest_salary)

Mean: 2522500.0
Median: 32500.0

Employee with highest salary:
Employee           D
Salary      10000000
Name: 3, dtype: object


# Section - D : Scenario-based Questions

### Q1. A student assessment dataset contains missing marks, duplicate student records, inconsistent department names such as IT, it, and I.T., and marks stored as text. Develop a cleaning sequence to standardize the data, handle missing values, remove duplicates, and convert marks into a suitable numeric type. State one risk of deleting missing values without investigating why they are missing. 

### Given Dataset: `student_assessment_dirty.csv`

The dataset contains the following records:

| Student_ID | Student_Name | Department | Assessment | Marks |
|------------|--------------|------------|------------|-------|
| S101 | Asha | IT | Python | 85 |
| S102 | Ravi | it | Python | 78 |
| S103 | Neha | I.T. | Python | 92 |
| S104 | Karan | HR | Python | |
| S105 | Meera | HR | Python | Absent |
| S106 | Arjun | Finance | Python | 0 |
| S107 | Priya | finance | Python | 67 |
| S108 | Rahul | IT | Python | 75 |
| S102 | Ravi | it | Python | 78 |
| S109 | Sneha | I.T. | Python | NA |
| S110 | Varun | HR | Python | 88 marks |
| S111 | Diya | Finance | Python | 91 |
| S112 | Aman | IT | Python | |
| S113 | Nisha | HR | Python | 64 |
| S114 | Kabir | Finance | Python | 105 |

### 1.Load the CSV file and inspect the dataset using head(), info(), and isnull().sum().


In [13]:
import pandas as pd
df = pd.read_csv("student_assessment_dirty.csv")
print("First five rows:")
print(df.head())
print("\nDataset information:")
df.info()
print("\nMissing values:")
print(df.isnull().sum())

First five rows:
  Student_ID Student_Name Department Assessment   Marks
0       S101         Asha         IT     Python      85
1       S102         Ravi         it     Python      78
2       S103         Neha       I.T.     Python      92
3       S104        Karan         HR     Python     NaN
4       S105        Meera         HR     Python  Absent

Dataset information:
<class 'pandas.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   Student_ID    15 non-null     str  
 1   Student_Name  15 non-null     str  
 2   Department    15 non-null     str  
 3   Assessment    15 non-null     str  
 4   Marks         12 non-null     str  
dtypes: str(5)
memory usage: 732.0 bytes

Missing values:
Student_ID      0
Student_Name    0
Department      0
Assessment      0
Marks           3
dtype: int64


### Explanation:
First, we load the CSV file using pd.read_csv(). Then, head() is used to view the first few rows, info() is used to check the structure and data types, and isnull().sum() is used to identify the number of missing values in each column. This inspection helps us understand the problems in the dataset before cleaning it

### 2.Standardize department names so that IT, it, and I.T. are represented consistently. Apply the same approach to Finance.


In [14]:
df["Department"] = (
    df["Department"]
    .str.strip()
    .str.upper()
    .replace({
        "I.T.": "IT",
        "IT": "IT",
        "FINANCE": "FINANCE",
        "HR": "HR"
    })
)
print(df["Department"].value_counts())

Department
IT         7
HR         4
FINANCE    4
Name: count, dtype: int64


### Explanation:
We standardize the department names by converting all names to uppercase and removing unnecessary dots. This makes IT, it, and I.T. consistent as IT, which helps avoid errors during data analysis.

### 3.Identify and remove exact duplicate student assessment records.

In [15]:
before = len(df)

df = df.drop_duplicates()

removed = before - len(df)

print("Records removed:", removed)

Records removed: 1


### Explanation:
We standardize the department names by converting all names to uppercase and removing unnecessary dots. This makes IT, it, and I.T. consistent as IT, which helps avoid errors during data analysis.

#### 4.Convert the Marks column into a numeric format, handling "88 marks", "Absent", "NA", and blank entries appropriately.


In [16]:
df["Marks"] = (
    df["Marks"]
    .astype(str)
    .str.extract(r"(\d+\.?\d*)")[0]
)

df["Marks"] = pd.to_numeric(df["Marks"], errors="coerce")

print(df["Marks"])

0      85.0
1      78.0
2      92.0
3       NaN
4       NaN
5       0.0
6      67.0
7      75.0
9       NaN
10     88.0
11     91.0
12      NaN
13     64.0
14    105.0
Name: Marks, dtype: float64


### Explanation:
The Marks column is converted into a numeric format by extracting the numbers from text values such as "88 marks". Values such as "Absent", "NA", and blank entries cannot be converted into marks, so they are treated as missing values (NaN).

### 5.Investigate missing marks and distinguish absent students from students whose marks have not been entered. State one risk of deleting missing values without investigating why they are missing.


In [18]:
print(df[df["Marks"].isna()])

   Student_ID Student_Name Department Assessment  Marks
3        S104        Karan         HR     Python    NaN
4        S105        Meera         HR     Python    NaN
9        S109        Sneha         IT     Python    NaN
12       S112         Aman         IT     Python    NaN


### Explanation:

The isna() function is used to identify missing values in the Marks column. The code filters and displays only the student records where marks are not available. This helps us investigate missing marks before deciding whether to remove them or replace them.

### 6.Identify marks outside the valid range of 0–100 and decide how they should be handled.


In [20]:
invalid = df[(df["Marks"] < 0) | (df["Marks"] > 100)]

print(invalid)

   Student_ID Student_Name Department Assessment  Marks
14       S114        Kabir    FINANCE     Python  105.0


### Explanation:

Identify marks below 0 or above 100, verify the original data, correct data-entry errors, and use NaN if the correct value cannot be determined.

### 7.Display the cleaned DataFrame and report the number of records removed, missing marks remaining, and invalid marks identified.



In [21]:
print(df)

print("Records removed:", removed)
print("Missing marks remaining:", df["Marks"].isna().sum())
print("Invalid marks identified:", len(invalid))

   Student_ID Student_Name Department Assessment  Marks
0        S101         Asha         IT     Python   85.0
1        S102         Ravi         IT     Python   78.0
2        S103         Neha         IT     Python   92.0
3        S104        Karan         HR     Python    NaN
4        S105        Meera         HR     Python    NaN
5        S106        Arjun    FINANCE     Python    0.0
6        S107        Priya    FINANCE     Python   67.0
7        S108        Rahul         IT     Python   75.0
9        S109        Sneha         IT     Python    NaN
10       S110        Varun         HR     Python   88.0
11       S111         Diya    FINANCE     Python   91.0
12       S112         Aman         IT     Python    NaN
13       S113        Nisha         HR     Python   64.0
14       S114        Kabir    FINANCE     Python  105.0
Records removed: 1
Missing marks remaining: 4
Invalid marks identified: 1


### Explanation:

The cleaned DataFrame shows the data after cleaning. Records removed tells how many rows were deleted. Missing marks remaining tells how many marks are still empty. Invalid marks identified tells how many marks were below 0 or above 100. This helps us check whether the data is clean and correct.


                                                       ◆  END OF THE ASSIGNMENT ◆